### PR TITLE
Add runtime configuration for minimum layer size for remote mount

### LINKED ISSUE #172: A runtime override for min-layer-size
**Is your feature request related to a problem? Please describe.**

`soci create --min-layer-size` let's users say "don't bother creating a zTOC for small layers, in which case the snapshotter will just pull the layer at launch time like the default snapshotter does. This is great for performance tuning a particular image, but it means that multiple SOCI indices have to be generated and pushed when running benchmarks to find the right size. 

**Describe the solution you'd like**

A more convenient solution would be `min_layer_size` option in  _snapshotterConfig_. If set, it would mean, "even if the SOCI index supplied a zTOC for the layer, if the layer is small, then ignore the zTOC and do the MountLocal() thing instead.

Another use case for this feature would be to allow administrators to override what's in the SOCI index when running on systems with a very slow or unreliable network connection to the remote registry. (To do this today, they would have to create and managed multiple  SOCI indices.)

### PR DESCRIPTION
**Pull request by sparr**
*Issue #, if available:*
closes #172 

*Description of changes:*
Adds an option to `soci-snapshotter-grpc/config.toml` to set a minimum size for remote mounting. Smaller layers will always be locally mounted.

*Testing performed:*
Ran automated test, and manual similar tests with other images and thresholds, with and without new config.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: snapshot/snapshot.go [hunk 1]
```diff
@@ -116,6 +120,12 @@ func AsynchronousRemove(config *SnapshotterConfig) error {
 	return nil
 }
 
+// MinLayerSize sets the smallest layer that will be mounted remotely.
+func MinLayerSize(config *SnapshotterConfig, minLayerSize int) error {
```
Comment (rdpsin) [inline snapshot/snapshot.go:124]: This should really be a function that returns `Opt`.

```
func WithMinLayerSize(minLayerSize int) Opt {
      return func(c *SnapshotterConfig) error {
               c.minLayerSize = minLayerSize
               return nil
      }
}
```

File: service/config.go [hunk 2]
```diff
@@ -48,6 +48,9 @@ type Config struct {
 
 	// ResolverConfig is config for resolving registries.
 	ResolverConfig `toml:"resolver"`
+
+	// MinLayerSize skips remote mounting of smaller layers
+	MinLayerSize int `toml:"min_layer_size"`
```
Comment (rdpsin) [inline service/config.go:53]: #345 adds `SnapshotterConfig` to `fs/config`. This field should be part of that.

np: How about renaming it to something like `MinLayerSizeOverride`, to distinguish it from the `MinLayerSize` in the CLI?

File: snapshot/snapshot.go [hunk 3]
```diff
@@ -268,37 +281,54 @@ func (o *snapshotter) Prepare(ctx context.Context, key, parent string, opts ...s
...
-			log.G(lCtx).WithField(remoteSnapshotLogKey, prepareFailed).
-				WithError(err).Warn("failed to prepare remote snapshot")
-			if !errors.Is(err, ErrNoZtoc) {
-				commonmetrics.IncOperationCount(commonmetrics.FuseMountFailureCount, digest.Digest(""))
+
+		localOverride := false
+		if strVal, ok := base.Labels["containerd.io/snapshot/remote/image.layers.size"]; ok {
+			if intVal, err := strconv.Atoi(strVal); err == nil {
+				if intVal <= o.minLayerSize {
+					localOverride = true
```
Comment (rdpsin) [inline snapshot/snapshot.go:297]: Should we emit a metric here?

File: service/config.go (hunk 2 above)
Comment (sbuckfelder) [reply service/config.go:53]: Heck might as well call it `MinLayerSizeConfigOverride` no reason not to be explicit

Comment (sparr) [reply service/config.go:53]: SnapshotterConfig is used for different purposes than snapshotterConfig. I think it would need to be in both. I'd be moving it out of the snapshotter struct, not out of the config for the main app.

File: snapshot/snapshot.go (hunk 1 above)
Comment (sparr) [reply snapshot/snapshot.go:124]: I discussed this with Kern. This function follows the model of AsynchronousRemove, which directly manipulates the config instead of returning a function that manipulates the config. This model is distinct from the WithFoo model used elsewhere in the project for options.
Making this change would require also changing AsynchronousRemove, and the handler that calls them.

File: snapshot/snapshot.go (hunk 3 above)
Comment (sparr) [reply snapshot/snapshot.go:297]: And, if so, which sort of metric?

File: service/config.go (hunk 2 above)
Comment (sparr) [reply service/config.go:53]: The CLI min layer size is for creating, not preparing/mouinting. This new option is not an override of that option, it applies at a different point in the workflow. If you use both, the larger one overrides the smaller, in either order.

File: snapshot/snapshot.go [hunk 4]
```diff
@@ -268,37 +281,54 @@ func (o *snapshotter) Prepare(ctx context.Context, key, parent string, opts ...s
...
 		lCtx := log.WithLogger(ctx, log.G(ctx).WithField("key", key).WithField("parent", parent))
-		if err := o.prepareRemoteSnapshot(lCtx, key, base.Labels); err != nil {
-			log.G(lCtx).WithField(remoteSnapshotLogKey, prepareFailed).
-				WithError(err).Warn("failed to prepare remote snapshot")
-			if !errors.Is(err, ErrNoZtoc) {
-				commonmetrics.IncOperationCount(commonmetrics.FuseMountFailureCount, digest.Digest(""))
+
+		localOverride := false
+		if strVal, ok := base.Labels["containerd.io/snapshot/remote/image.layers.size"]; ok {
+			if intVal, err := strconv.Atoi(strVal); err == nil {
```
Comment (Kern--) [inline snapshot/snapshot.go:295]: We should log an error if we can't convert to an int.

File: snapshot/snapshot.go [hunk 5]
```diff
@@ -268,37 +281,54 @@ func (o *snapshotter) Prepare(ctx context.Context, key, parent string, opts ...s
...
 		//       log is used by tests in this project.
 		lCtx := log.WithLogger(ctx, log.G(ctx).WithField("key", key).WithField("parent", parent))
-		if err := o.prepareRemoteSnapshot(lCtx, key, base.Labels); err != nil {
-			log.G(lCtx).WithField(remoteSnapshotLogKey, prepareFailed).
-				WithError(err).Warn("failed to prepare remote snapshot")
-			if !errors.Is(err, ErrNoZtoc) {
-				commonmetrics.IncOperationCount(commonmetrics.FuseMountFailureCount, digest.Digest(""))
+
+		localOverride := false
+		if strVal, ok := base.Labels["containerd.io/snapshot/remote/image.layers.size"]; ok {
```
Comment (Kern--) [inline snapshot/snapshot.go:294]: nit: can we export this field and use that instead? 

https://github.com/awslabs/soci-snapshotter/blob/main/fs/source/source.go#L95

File: snapshot/snapshot.go [hunk 6]
```diff
@@ -268,37 +281,54 @@ func (o *snapshotter) Prepare(ctx context.Context, key, parent string, opts ...s
...
+			if intVal, err := strconv.Atoi(strVal); err == nil {
+				if intVal <= o.minLayerSize {
```
Comment (Kern--) [inline snapshot/snapshot.go:296]: nit: I think this should be strictly less than. If I say "Min layer size is 10MiB" I would expect a file that is exactly 10MiB to be remotely mounted. TBF, I wouldn't be surprised if we never hit this case, but it seems esrier to parse to me.

File: snapshot/snapshot.go [hunk 7]
```diff
@@ -682,10 +712,12 @@ func (o *snapshotter) mounts(ctx context.Context, s storage.Snapshot, checkKey s
 
 }
 
+// Produces a file path like "{snapshotter.root}/snapshots/{id}/fs"
```
Comment (Kern--) [inline snapshot/snapshot.go:715]: nit: I think we should follow the normal godoc format and start with the name of the thing we're documenting. It makes it easier to export the function later.

File: snapshot/snapshot.go (hunk 1 above)
Comment (Kern--) [reply snapshot/snapshot.go:124]: `AsynchronousRemove` is already an `Opt`. Personally, I think `AsynchronousRemove` should be renamed to `WithAsynchronousRemove` and we should use @rdpsin's suggestion here.

**Review by Kern-- — DISMISSED** [review]

**Comment by Kern--** [comment]
LGTM pending tests.

**Comment by rdpsin** [comment]
LGTM. Can we have an integration test for this?

File: integration/pull_test.go [hunk 8]
```diff
@@ -49,6 +49,14 @@ import (
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
+const (
+	minLayerSize = "20000000"
```
Comment (Kern--) [inline integration/pull_test.go:53]: nit for follow up: the constant for `minLayerSize` and expected FUSE mounts means that this test could fail unexpectedly if the number or size of layers in the latest image change. We can partially solve this with pinning the version we test  against, but it would still be annoying if tests failed when upgrading the pinned version.

Instead of making these constants, it would be nice if we can choose a size that is exactly the median size of layers in the image and then calculate how many fuse mounts we expect. That way we can be sure it won't start failing after an update and we can test the edge case where min layer size is exactly the size of a layer. I think we can do all this by calling `rebootContainerd` after creating the SOCI index to remove the local state of containerd/the snapshotter.

Can you create an issue for that follow up?

**Review by Kern-- — APPROVED** [review]

File: integration/pull_test.go (hunk 8 above)
Comment (sparr) [reply integration/pull_test.go:53]: https://github.com/awslabs/soci-snapshotter/issues/373

File: integration/pull_test.go [hunk 9]
```diff
@@ -845,3 +853,42 @@ func TestRpullImageThenRemove(t *testing.T) {
...
+		t.Fatalf("invalid soci index from digest %s: %v", indexDigest, rawJSON)
+	}
+
+	if len(sociIndex.Blobs) == 0 {
+		t.Fatalf("soci index %s contains 0 blobs, invalidating this test", indexDigest)
+	}
+
+	sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)
+
+	checkFuseMounts(t, sh, 3) // default behavior produces 10 Fuse mounts
```
Comment (rdpsin) [inline integration/pull_test.go:893]: Is this right? The default behavior should produce 3 mounts, since the defaultMinLayerSize is set to 10MiB.

File: integration/pull_test.go [hunk 10]
```diff
@@ -845,3 +853,42 @@ func TestRpullImageThenRemove(t *testing.T) {
...
+	sh.
+		X("update-ca-certificates").
+		Retry(100, "nerdctl", "login", "-u", regConfig.user, "-p", regConfig.pass, regConfig.host)
+
+	rebootContainerd(t, sh, "", "")
+
+	containerImage := rabbitmqImage
+
+	copyImage(sh, dockerhub(containerImage), regConfig.mirror(containerImage))
+	indexDigest := buildIndex(sh, regConfig.mirror(containerImage))
```
Comment (rdpsin) [inline integration/pull_test.go:879]: We should probably pass `WithMinLayerSize(0)` to ensure that ztocs for all layers are being built.

File: integration/pull_test.go (hunk 9 above)
Comment (sparr) [reply integration/pull_test.go:893]: I did run it both ways, and it produced 10 without a runtime min layer size set. I'll see if the creation-time min layer size is getting overridden in some way, or what other explanation there is for my results.

File: integration/pull_test.go (hunk 10 above)
Comment (sparr) [reply integration/pull_test.go:879]: That feels like a different case that might deserve its own test. What I'd like to be doing here is comparing to the default behavior.

Comment (Kern--) [reply integration/pull_test.go:879]: https://github.com/awslabs/soci-snapshotter/pull/363 just changed the default behavior for tests from 0 to `defaultMinLayerSize`. `WithMinLayerSize(0)` is probably equivalent to what you actually tested against.

File: integration/pull_test.go (hunk 9 above)
Comment (sparr) [reply integration/pull_test.go:893]: See other thread below.

File: integration/pull_test.go (hunk 10 above)
Comment (sparr) [reply integration/pull_test.go:879]: Good catch! That would explain the discrepancy.

Comment (sparr) [reply integration/pull_test.go]: I believe that I've addressed this.

**Review by rdpsin — APPROVED** [review]

**Review by Kern-- — APPROVED** [review]
